File: sort_algorithms/bucket_sort.py

# Bucket Sort
# Python
# Porter L

import insertion_sort as iSort  
              
def bucketSort(Arr):
    max_value = max(Arr)
    size = max_value/len(Arr)
    buckets_list = []


    for x in range(len(Arr)): # create buckets
        buckets_list.append([]) 

    for y in range(len(Arr)):
        z = int (Arr[y] / size)
        if (z != len (Arr)):
            buckets_list[z].append(Arr[y])
        else:
            buckets_list[len(Arr) - 1].append(Arr[y])

    for z in range(len(Arr)): # use previously implemented insertion sort 
        iSort.insertionSort(buckets_list[z])
            
    Arr_out = []
    for x in range(len (Arr)): # combine buckets into final output
        Arr_out = Arr_out + buckets_list[x]
    return Arr_out

# Example
if __name__=='__main__': 
    test_array = [2,7,22,71,5,3,11,9,7,5,4,3,56,3,23,43,97,3]
    print(test_array) # unsorted
    test_array = bucketSort(test_array)
    print(test_array) # sorted

File: sort_algorithms/javascript/bucket_sort.js
// Bucket Sort
// Javascript
// Porter L

function bucketSort(Arr){
    console.log('bucket sort');
    var max_value = Math.max(...Arr);
    var size = max_value / Arr.length;
    var buckets_list = [];

    for (x = 0; x < Arr.length; x++){
        buckets_list.push([]);
    }

    for (y = 0; y < Arr.length; y++){
        var z = Math.floor(Arr[y] / size);
        if (z != Arr.length){
            buckets_list[z].push(Arr[y]);
        }
        else{
            buckets_list[Arr.length - 1].push(Arr[y]);
        }
    }

    for (z = 0; z < Arr.length; z++){
        insertionSort(buckets_list[z]);
    }

    var Arr_out = [];
    for (x = 0; x < Arr.length; x++){
        Arr_out = Arr_out.concat(...buckets_list[x]);
    }
    return Arr_out
}